src/system/SystemPacketBuffer.cpp

```diff
@@ -508,45 +508,63 @@ void PacketBuffer::AddRef()
 
 PacketBufferHandle PacketBufferHandle::New(size_t aAvailableSize, uint16_t aReservedSize)
 {
-    // Adding three 16-bit-int sized numbers together will never overflow
-    // assuming int is at least 32 bits.
+    // Ensure that SIZE_MAX is at least 32 bits
     static_assert(INT_MAX >= INT32_MAX, "int is not big enough");
+    static_assert(SIZE_MAX >= INT_MAX, "Our additions may not fit in size_t");
+
+    // Sanity check for kStructureSize and putting an upper bound.
     static_assert(PacketBuffer::kStructureSize == sizeof(PacketBuffer), "PacketBuffer size mismatch");
-    static_assert(PacketBuffer::kStructureSize < UINT16_MAX, "Check for overflow more carefully");
-    static_assert(SIZE_MAX >= INT_MAX, "Our additions might not fit in size_t");
-    static_assert(PacketBuffer::kMaxSizeWithoutReserve <= UINT32_MAX, "PacketBuffer may have size not fitting uint32_t");
+    static_assert(PacketBuffer::kStructureSize < UINT16_MAX, "kStructureSize should not exceed UINT16_MAX");
+    static_assert(PacketBuffer::kMaxSizeWithoutReserve <= UINT16_MAX, "PacketBuffer may have size not fitting uint16_t");
+
+    // Have a bounds check for aAvailableSize to ensure that
+    // the next overflow check is done correctly.
+    if (aAvailableSize > UINT32_MAX)
+    {
+        ChipLogError(chipSystemLayer, "PacketBuffer: AvailableSize of a buffer cannot exceed UINT32_MAX");
+        return PacketBufferHandle();
+    }
+
+    // Cast all to uint64_t and add to consider the widest possible result.
```

Comment:
```suggestion
    // Cast all to uint64_t and add.  This can't overflow, because we have ensured that the maximal sum here is
    // UINT32_MAX + UINT16_MAX + UINT16_MAX, which always fits in a 64-bit unsigned integer. 
```